services/activitylog/pkg/service/service.go

```diff
@@ -265,70 +376,113 @@ func (a *ActivitylogService) activities(rid *provider.ResourceId) ([]RawActivity
 	var (
 		info  *provider.ResourceInfo
-		err   error
 		depth int
 		ref   = initRef
 	)
 	for {
-		info, err = getResource(ref)
-		if err != nil {
-			return fmt.Errorf("could not get resource info: %w", err)
+		id := ref.GetResourceId()
+		if ref.Path != "" {
+			// Path based reference, we need to resolve the resource id
+			info, err := getResource(ref)
+			if err != nil {
+				return fmt.Errorf("could not get resource info: %w", err)
+			}
+			id = info.GetId()
 		}
-
-		if err := a.storeActivity(storagespace.FormatResourceID(info.GetId()), eventID, depth, timestamp); err != nil {
-			return fmt.Errorf("could not store activity: %w", err)
+		if id == nil {
+			return fmt.Errorf("resource id is required")
 		}
 
-		if info != nil && utils.IsSpaceRoot(info) {
+		key := storagespace.FormatResourceID(id)
+		_, span := a.tracer.Start(ctx, "queueStoreActivity")
+		a.debouncer.Debounce(key, RawActivity{
+			EventID:   eventID,
+			Depth:     depth,
+			Timestamp: timestamp,
+		})
+		span.End()
+
+		if id.OpaqueId == id.SpaceId {
+			// we are at the root of the space, no need to go further
 			return nil
 		}
 
+		// check if parent id is cached
+		// parent id is cached in the format <storageid>$<spaceid>!<resourceid>
+		// if it is not cached, get the resource info and cache it
+		if parentId == nil {
+			if v, err := a.parentIdCache.Get(key); err != nil {
+				_, span = a.tracer.Start(ctx, "getResource")
+				info, err = getResource(ref)
+				span.End()
+				if err != nil || info.GetParentId() == nil || info.GetParentId().GetOpaqueId() == "" {
+					return fmt.Errorf("could not get parent id: %w", err)
+				}
+				parentId = info.GetParentId()
+				a.parentIdCache.Set(key, parentId)
+			} else {
+				parentId = v.(*provider.ResourceId)
+			}
+		} else {
+			a.log.Debug().Msg("parent id is cached")
+		}
+
 		depth++
-		ref = &provider.Reference{ResourceId: info.GetParentId()}
+		ref = &provider.Reference{ResourceId: parentId}
+		parentId = nil // reset parent id so it's not reused in the next iteration
 	}
 }
 
-func (a *ActivitylogService) storeActivity(resourceID string, eventID string, depth int, timestamp time.Time) error {
+func (a *ActivitylogService) storeActivity(resourceID string, activities []RawActivity) error {
 	a.lock.Lock()
 	defer a.lock.Unlock()
 
+	ctx, span := a.tracer.Start(context.Background(), "storeActivity")
+	defer span.End()
+	_, subspan := a.tracer.Start(ctx, "store.Read")
 	records, err := a.store.Read(resourceID)
 	if err != nil && err != microstore.ErrNotFound {
 		return err
 	}
+	subspan.End()
 
-	var activities []RawActivity
+	_, subspan = a.tracer.Start(ctx, "Unmarshal")
+	var existingActivities []RawActivity
 	if len(records) > 0 {
-		if err := json.Unmarshal(records[0].Value, &activities); err != nil {
-			return err
+		if err := msgpack.Unmarshal(records[0].Value, &existingActivities); err != nil {
```

Comment:
Nit: Hm, somehow I think we should have a debug log statement here when falling back json. (Feel free to ignore me here)